File: go-spider/internal/security/outbound_policy.go
package security

import (
	"context"
	"time"
)

func preDNSDecision(normalized NormalizedURL, policy OutboundPolicy, previous *NormalizedURL) *PolicyDecision {
	if normalized.Scheme != "http" && normalized.Scheme != "https" {
		return &PolicyDecision{Allowed: false, ReasonCode: "scheme_not_allowed", Normalized: &normalized, Addresses: []string{}}
	}
	if normalized.Scheme == "http" && !policy.AllowedHTTP() {
		return &PolicyDecision{Allowed: false, ReasonCode: "http_not_allowed", Normalized: &normalized, Addresses: []string{}}
	}

	if previous == nil {
		if !policy.AllowsHost(normalized.Host) {
			return &PolicyDecision{Allowed: false, ReasonCode: "host_not_allowed", Normalized: &normalized, Addresses: []string{}}
		}
	} else {
		if !policy.AllowsHost(normalized.Host) {
			return &PolicyDecision{Allowed: false, ReasonCode: "redirect_host_not_allowed", Normalized: &normalized, Addresses: []string{}}
		}
	}

	if !policy.AllowsPort(normalized.Scheme, normalized.Port) {
		return &PolicyDecision{Allowed: false, ReasonCode: "port_not_allowed", Normalized: &normalized, Addresses: []string{}}
	}
	if previous != nil && previous.Scheme == "https" && normalized.Scheme == "http" {
		return &PolicyDecision{Allowed: false, ReasonCode: "https_downgrade", Normalized: &normalized, Addresses: []string{}}
	}
	return nil
}

func postDNSDecision(normalized NormalizedURL, dns DNSValidationResult) PolicyDecision {
	if !dns.Allowed {
		reason := dns.ReasonCode
		if reason == "" {
			reason = "ip_not_allowed"
		}
		return PolicyDecision{Allowed: false, ReasonCode: reason, Normalized: &normalized, Addresses: dns.Addresses}
	}
	return PolicyDecision{Allowed: true, Normalized: &normalized, Addresses: dns.Addresses}
}

func EvaluatePolicy(normalized NormalizedURL, dns DNSValidationResult, policy OutboundPolicy, previous *NormalizedURL) PolicyDecision {
	if pre := preDNSDecision(normalized, policy, previous); pre != nil {
		return *pre
	}
	return postDNSDecision(normalized, dns)
}

func DecidePolicy(raw string, policy OutboundPolicy, resolver Resolver) (PolicyDecision, error) {
	normalized, err := NormalizeOutboundURL(raw)
	if err != nil {
		return PolicyDecision{Allowed: false, ReasonCode: reasonCode(err), Normalized: nil}, nil
	}
	if pre := preDNSDecision(normalized, policy, nil); pre != nil {
		return *pre, nil
	}
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	dns, err := ResolveAndValidate(ctx, normalized.Host, resolver, policy.MaxDNSAddresses())
	if err != nil {
		return PolicyDecision{Allowed: false, ReasonCode: reasonCode(err), Normalized: &normalized}, nil
	}
	return postDNSDecision(normalized, dns), nil
}

func DecideRedirect(previousRaw, targetRaw string, policy OutboundPolicy, resolver Resolver) (PolicyDecision, error) {
	previous, err := NormalizeOutboundURL(previousRaw)
	if err != nil {
		return PolicyDecision{Allowed: false, ReasonCode: reasonCode(err), Normalized: nil}, nil
	}
	target, err := NormalizeOutboundURL(targetRaw)
	if err != nil {
		return PolicyDecision{Allowed: false, ReasonCode: reasonCode(err), Normalized: nil}, nil
	}
	if pre := preDNSDecision(target, policy, &previous); pre != nil {
		return *pre, nil
	}
	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
	defer cancel()
	dns, err := ResolveAndValidate(ctx, target.Host, resolver, policy.MaxDNSAddresses())
	if err != nil {
		return PolicyDecision{Allowed: false, ReasonCode: reasonCode(err), Normalized: &target}, nil
	}
	return postDNSDecision(target, dns), nil
}

func reasonCode(err error) string {
	if e, ok := err.(*URLNormalizationError); ok {
		return e.ReasonCode
	}
	if e, ok := err.(*DNSValidationError); ok {
		return e.ReasonCode
	}
	return "invalid_url"
}

File: crawler/security/outbound_policy.py
"""Pure outbound policy decisions for TASK-022B."""

from __future__ import annotations

from crawler.security.dns_policy import DNSValidationError, Resolver, resolve_and_validate
from crawler.security.models import DNSValidationResult, NormalizedURL, OutboundPolicy, PolicyDecision
from crawler.security.url_normalizer import URLNormalizationError, normalize_outbound_url


def default_policy() -> OutboundPolicy:
    return OutboundPolicy.default()


def _pre_dns_decision(
    normalized: NormalizedURL,
    policy: OutboundPolicy,
    previous: NormalizedURL | None = None,
) -> PolicyDecision | None:
    if normalized.scheme not in ("http", "https"):
        return PolicyDecision(False, "scheme_not_allowed", normalized, ())
    if normalized.scheme == "http" and not policy.allowed_http:
        return PolicyDecision(False, "http_not_allowed", normalized, ())

    allowed_hosts = {host.lower() for host in policy.allowed_hosts}
    if previous is None:
        if normalized.host not in allowed_hosts:
            return PolicyDecision(False, "host_not_allowed", normalized, ())
    else:
        if normalized.host not in allowed_hosts:
            return PolicyDecision(False, "redirect_host_not_allowed", normalized, ())

    allowed_ports = set(policy.allowed_ports_by_scheme.get(normalized.scheme, frozenset()))
    if normalized.port not in allowed_ports:
        return PolicyDecision(False, "port_not_allowed", normalized, ())

    if previous is not None and previous.scheme == "https" and normalized.scheme == "http":
        return PolicyDecision(False, "https_downgrade", normalized, ())
    return None


def _post_dns_decision(
    normalized: NormalizedURL,
    dns: DNSValidationResult,
    policy: OutboundPolicy,
) -> PolicyDecision:
    if not dns.allowed:
        return PolicyDecision(False, dns.reason_code or "ip_not_allowed", normalized, dns.addresses)
    return PolicyDecision(True, "", normalized, dns.addresses)


def evaluate(
    normalized: NormalizedURL,
    dns: DNSValidationResult,
    policy: OutboundPolicy,
    previous: NormalizedURL | None = None,
) -> PolicyDecision:
    pre = _pre_dns_decision(normalized, policy, previous=previous)
    if pre is not None:
        return pre
    return _post_dns_decision(normalized, dns, policy)


def decide(
    raw_url: str,
    policy: OutboundPolicy,
    *,
    resolver: Resolver | None = None,
    timeout_seconds: float = 5.0,
    previous_raw: str | None = None,
) -> PolicyDecision:
    try:
        normalized = normalize_outbound_url(raw_url)
    except URLNormalizationError as exc:
        return PolicyDecision(False, exc.reason_code, None, ())

    previous: NormalizedURL | None = None
    if previous_raw is not None:
        try:
            previous = normalize_outbound_url(previous_raw)
        except URLNormalizationError as exc:
            return PolicyDecision(False, exc.reason_code, None, ())

    pre = _pre_dns_decision(normalized, policy, previous=previous)
    if pre is not None:
        return pre

    if resolver is None and not normalized.is_ip_literal:
        raise ValueError("resolver is required for hostname URLs")
    try:
        dns = resolve_and_validate(normalized.host, resolver, timeout_seconds=timeout_seconds)
    except DNSValidationError as exc:
        return PolicyDecision(False, exc.reason_code, normalized, ())

    return _post_dns_decision(normalized, dns, policy)
